pmm/queries.go

```diff
@@ -18,22 +18,367 @@
+			return nil, err
+		}
+	} else {
+		port = consulSvc.Port
+		// Ensure qan-agent is started if service exists, otherwise it won't be enabled for QAN.
+		if err := startService(fmt.Sprintf("pmm-%s-queries-%d", q.Name(), port)); err != nil {
+			return nil, err
+		}
+	}
+
+	// Start QAN by associating instance with agent.
+	qanConfig := q.Config()
+	qanConfig.UUID = instance.UUID
+	qanConfig.Interval = 60
+	if err := a.startQAN(agentID, qanConfig); err != nil {
+		return nil, err
+	}
+
+	tags := []string{
+		fmt.Sprintf("alias_%s", a.ServiceName),
+	}
+	// For existing service, we append a new alias_ tag.
+	if consulSvc != nil {
+		tags = append(consulSvc.Tags, tags...)
+	}
+
+	// Add service to Consul.
+	serviceID := fmt.Sprintf("%s-%d", serviceType, port)
+	srv := consul.AgentService{
+		ID:      serviceID,
+		Service: serviceType,
+		Tags:    tags,
+		Port:    port,
+	}
+	reg := consul.CatalogRegistration{
+		Node:    a.Config.ClientName,
+		Address: a.Config.ClientAddress,
+		Service: &srv,
+	}
+	if _, err := a.consulAPI.Catalog().Register(&reg, nil); err != nil {
+		return nil, err
+	}
+
+	// Add info to Consul KV.
+	d := &consul.KVPair{
+		Key:   fmt.Sprintf("%s/%s/%s/dsn", a.Config.ClientName, serviceID, a.ServiceName),
+		Value: []byte(utils.SanitizeDSN(info.DSN)),
+	}
+	a.consulAPI.KV().Put(d, nil)
+	d = &consul.KVPair{
+		Key:   fmt.Sprintf("%s/%s/%s/qan_%s_uuid", a.Config.ClientName, serviceID, a.ServiceName, q.Name()),
+		Value: []byte(instance.UUID),
+	}
+	a.consulAPI.KV().Put(d, nil)
+
+	return info, nil
+}
+
+// RemoveMySQLQueries remove mysql instance from QAN.
+func (a *Admin) RemoveQueries(name string) error {
+	serviceType := fmt.Sprintf("%s:queries", name)
+
+	// Check if we have this service on Consul.
+	consulSvc, err := a.getConsulService(serviceType, a.ServiceName)
+	if err != nil {
+		return err
+	}
+	if consulSvc == nil {
+		return ErrNoService
+	}
+
+	// Ensure qan-agent is started, otherwise it will be an error to stop QAN.
+	if err := startService(fmt.Sprintf("pmm-%s-queries-%d", name, consulSvc.Port)); err != nil {
+		return err
+	}
+
+	// Get UUID of MySQL instance the agent is monitoring from KV.
+	key := fmt.Sprintf("%s/%s/%s/qan_%s_uuid", a.Config.ClientName, consulSvc.ID, a.ServiceName, name)
+	data, _, err := a.consulAPI.KV().Get(key, nil)
+	if err != nil {
+		return err
+	}
+	if data == nil {
+		return fmt.Errorf("can't get key %s", key)
+	}
+	uuid := string(data.Value)
+
+	// Stop QAN for this instance on the local agent.
+	agentConfigFile := fmt.Sprintf("%s/config/agent.conf", AgentBaseDir)
+	agentID, err := getAgentID(agentConfigFile)
+	if err != nil {
+		return err
+	}
+	if err := a.stopQAN(agentID, uuid); err != nil {
+		return err
+	}
+
+	// Delete instance.
+	if err := a.deleteInstance(uuid); err != nil {
+		return err
+	}
+
+	prefix := fmt.Sprintf("%s/%s/%s/", a.Config.ClientName, consulSvc.ID, a.ServiceName)
+	a.consulAPI.KV().DeleteTree(prefix, nil)
```

Comment:
Error return value of `(*github.com/percona/pmm-client/vendor/github.com/hashicorp/consul/api.KV).DeleteTree` is not checked